probe-rs/src/probe/daplink/tools.rs

```diff
@@ -175,53 +173,51 @@ pub fn open_device_from_selector(
                 && sn_str == selector.serial_number
+                && get_daplink_info(&device).is_some()
             {
                 // If the VID, PID, and potentially SN all match,
+                // and the device is a valid CMSIS-DAP probe,
                 // attempt to open the device in v2 mode.
                 if let Some(device) = open_v2_device(device) {
                     return Ok(device);
-                } else {
-                    // If we are here the device didn't support v2, try using hidapi to open in v1 mode.
-                    // If this doesn't work we give up and return None.
-                    log::debug!(
-                        "Attempting to open {:04x}:{:04x} in CMSIS-DAP v1 mode",
-                        selector.vendor_id,
-                        selector.product_id
-                    );
-
-                    return match sn_str {
-                        Some(sn) => hidapi::HidApi::new().and_then(|api| {
-                            api.open_serial(selector.vendor_id, selector.product_id, &sn)
-                        }),
-                        None => hidapi::HidApi::new()
-                            .and_then(|api| api.open(selector.vendor_id, selector.product_id)),
-                    }
-                    .map(|device| DAPLinkDevice::V1(device))
-                    .map_err(From::from);
                 }
             }
         }
+    }
 
-        // No device was found matching the credentials or the matching device could not be opened.
-        Err(ProbeCreationError::NotFound)
-    } else {
-        // If rusb failed or the device didn't support v2, try using hidapi to open in v1 mode.
-        // If this doesn't work we give up and return None.
-        let vid = selector.vendor_id;
-        let pid = selector.product_id;
-        let sn = &selector.serial_number;
-
-        log::debug!(
-            "Attempting to open {:04x}:{:04x} in CMSIS-DAP v1 mode",
-            vid,
-            pid
-        );
-
-        return match sn {
-            Some(sn) => hidapi::HidApi::new().and_then(|api| api.open_serial(vid, pid, &sn)),
-            None => hidapi::HidApi::new().and_then(|api| api.open(vid, pid)),
+    // If rusb failed or the device didn't support v2, try using hidapi to open in v1 mode.
+    let vid = selector.vendor_id;
+    let pid = selector.product_id;
+    let sn = &selector.serial_number;
+
+    log::debug!(
+        "Attempting to open {:04x}:{:04x} in CMSIS-DAP v1 mode",
+        vid,
+        pid
+    );
+
+    // Attempt to open provided VID/PID/SN with hidapi
+    let hid_device = match sn {
+        Some(sn) => hidapi::HidApi::new().and_then(|api| api.open_serial(vid, pid, &sn)),
+        None => hidapi::HidApi::new().and_then(|api| api.open(vid, pid)),
+    };
+
+    match hid_device {
+        Ok(device) => {
+            match device.get_product_string() {
+                Ok(Some(s)) if s.contains("CMSIS-DAP") => Ok(DAPLinkDevice::V1(device)),
```

Comment:
Couldn't we use the `is_daplink` function here too? That would prevent weird inconsistency errors in the future if anything changes.

Reply:
Annoyingly no; `get_daplink_hid_info()` requires a `&hidapi::DeviceInfo` which comes from `HidApi::device_list()`; if you have a VID and PID and you've opened a `hidapi::Device`, you can't get a `DeviceInfo` from it, but you _can_ still read the product string to check it.

Reply:
Ok, so I think we are save to leave this for now, to get this merged and maybe change this in the future.

Reply:
I think the neatest thing would be if each probe's `open_device_from_selector()` only gets called with appropriate selectors, so it doesn't have to do this slightly confusing dance of returning NotFound for most errors. Instead, probe-rs could list all discovered probes using the normal `list_*_devices()` and then locate the selector inside that list, _then_ call `open_device_from_selector()` from the appropriate probe module only.

But like you say we don't need to fix it right now.